### ISSUE TITLE
Implement the Python standard library

### ISSUE THREAD

**Issue by freakboy3742** [pybee/voc#74]
Python has a sizeable standard library. It is this standard library the provides a lot of the power of Python as a language.

Much of this standard library is implemented in Python, and could be converted into a Jar file using VOC. VOC's build system should be modified to perform this conversion.

Other parts of the standard library are implemented in C, and will require re-implementation in Java.

Even if there is a pure-python implementation, it may be desirable to re-implement a standard module in Java for efficiency purposes.

This may be dependent on completion of #36 and #38, as you won't be able to use the standard library unless the builtin data types are available.

**Comment by freakboy3742** [comment]
Stubs of the sys and time modules already exist; they can be used as a template when adding new modules.

**Comment by freakboy3742** [comment]
See also [Ouroborous](https://github.com/pybee/ouroborous), a related project to develop a standalone, pure python version of the standard library. Ideally, it should be possible to use Ouroboros to generate the builtins and standard library.

**Comment by souravsingh** [comment]
@freakboy3742 I would like to work in the issue

**Comment by freakboy3742** [comment]
@souravsingh Fantastic! Do you need any pointers or suggestions, or do you feel confident tackling this on your own?

**Comment by souravsingh** [comment]
@freakboy3742 I just wanted to ask- We have to create Python library similar to Ouroboros, right?

**Comment by freakboy3742** [comment]
Ouroboros aims to be a full, pure-python version of the Python standard library (or, as much as is plausible), with a clearly delineated interface where platform-specific modules are required (e.g., to make system calls). 

I know this ticket is on VOC; but the ideal contribution would be to improve Ouroboros, because then *any* Python implementation can use it. The VOC component will then be just the Java implementation of the system native portions necessary to support Ouroboros on Java.

At some point, we'll need to find a way to integrate Ouroboros into the VOC build chain. 

So - as for what to do? Pick a standard library module that doesn't currently have an implementation in Ouroboros, and implement it. 

Two places worth starting: datetime and httplib. Both of these have pure Python components; but they also have native components. Make sure the full library is available.

The code that is in Ouroboros is using Python 3.4 as the reference implementation, so try to keep the APIs aligned with that version.

**Comment by amrinderbrar95** [comment]
I would like to attempt to solve this issue.

**Comment by abhinavralhan** [comment]
hey @freakboy3742 , is this issue still open?

**Comment by freakboy3742** [comment]
@abhinavralhan Yes - there's still plenty to implement in the standard library.

**Comment by abhinavralhan** [comment]
@freakboy3742 I would totally like to help. This would be my first contribution. Could you give me some pointers to get started?

**Comment by freakboy3742** [comment]
@abhinavralhan If this is your first contribution, I'd suggest starting somewhere else that will be a little easier. Our [getting started guide](https://pybee.org/contributing/how/first-time/what/voc/) for first time contributors lays out a suggested first contribution; once you've done that first exercise, more advanced issues like this one will make a lot more sense.

**Comment by hemanthsavasere** [comment]
Hi @freakboy3742,

Is this bug assigned to anyone ?

thanks,

**Comment by freakboy3742** [comment]
@hemanthsavasere This isn't a bug that is going to be assigned to a single person - there's a *huge* amount of work that needs to be done. This ticket is just for coordination purposes, so we don't end up with multiple people working on the same part. 

So - pick a part of the standard library (i.e., a specific module, like `collections` or `httplib`), drop a comment here that says what you're doing, and implement it!

**Comment by akarshsingh9** [comment]
Would like to get started. Where to start from?

**Comment by prithvisingh18** [comment]
Hi , @freakboy3742 I have started working on `time` library, as most of it is not implemented. 
After completing one of the function, I will put a Pull Request(For you to review).
By the way, I am a First Time Contribute. 
Do I need to put an comment on #36 also, as requested in the starting guide.

**Comment by evalevanto** [comment]
I definitely want to help. Checking the guide :)

**Comment by amaarahj** [comment]
Hi @freakboy3742 I am Software Engineering student with a class assignment in which I have to contribute to an open source project. I am interested in contributing to VOC. I went through the tutorial and first timer guides. I have some experience with Python and Java.

I am looking to get started with some functions in the time.java library: gmtime(), localtime(), mktime() and strptime(). I am planning for 3-4 hours of design, development, testing and rework for each function. I wanted to get you opinion on my plan given the complexity of these functions.

Thank you

**Comment by freakboy3742** [comment]
@amaarahj It's almost impossible for me to judge. I don't know you, your skills, or your experience level; as such, it's impossible for me to judge whether your estimates will be accurate. 

However, the four methods you've proposed are reasonably standalone, and sound like a reasonable target for a short university assignment.

**Comment by freakboy3742** [comment]
@gs-niteesh Absolutely - go right ahead!

**Comment by gs-niteesh** [comment]
Just having few issues. Do we have an exec function? and there are few functions like iskeyword which need to be implemented should I just continue with the implementation of namedtuple or implement the required one's?

**Comment by freakboy3742** [comment]
@gs-niteesh Yes - there are a lot of features missing. That's the point - there's a lot of work that needs to be done. 

NamedTuple in particular is a complex piece of code - and its native implementation in Python is a bit of a hack (as it is based on exec). It may turn out to be easier to implement in Java directly, rather than trying to execute the Python implementation.

**Comment by gs-niteesh** [comment]
Sorry for asking so many questions. But I don't understand what do you mean by implement it in java directly, can you explain it in brief.

**Comment by gs-niteesh** [comment]
Do you have any video or doc that explains VOC in detail. In detail I mean the internals and how things work.

**Comment by dhuberi1** [comment]
Hi - I am interested in working on this issue but i just wanted to clarify it is still open?

**Comment by akamanzi** [comment]
@freakboy3742, I would also be interesting in contributing, it is my first time. would love to know where i can start. let me know what is still open for first-timers.
Thanks

**Comment by aritramullick** [comment]
I would love to contribute. However, I am a  first timer so I would love to get some advice on how to go about starting this!

**Comment by freakboy3742** [comment]
@aritramullick At this point, I'd advise against starting with VOC as a project. We're in the process of removing it from use as part of our stack. If you're looking for project to contribute, I'd recommend [Briefcase](https://github.com/beeware/briefcase). Run through the [BeeWare Tutorial](https://docs.beeware.org) to see Briefcase in action, then see if there's a ticket in the Briefcase ticket tracker that appeals to you.
